src/main/java/net/snowflake/ingest/streaming/internal/LiteralQuoteUtils.java

```diff
@@ -0,0 +1,170 @@
+package net.snowflake.ingest.streaming.internal;
+
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.regex.Pattern;
+import net.snowflake.ingest.utils.Utils;
+
+/**
+ * Util class to normalise literals to match server side metadata.
+ *
+ * <p>Note: The methods in this class have to be kept in sync with the respective methods on server
+ * side.
+ */
+class LiteralQuoteUtils {
+  private static final String[] reservedKeywords = {
+    // ANSI reserved KW
+    "ALL",
+    "ALTER",
+    "AND",
+    "ANY",
+    "AS",
+    "BETWEEN",
+    "BY",
+    "CHECK",
+    "COLUMN",
+    "CONNECT",
+    "CREATE",
+    "CURRENT",
+    "DELETE",
+    "DISTINCT",
+    "DROP",
+    "ELSE",
+    "EXISTS",
+    "FOLLOWING",
+    "FOR",
+    "FROM",
+    "GRANT",
+    "GROUP",
+    "HAVING",
+    "IN",
+    "INSERT",
+    "INTERSECT",
+    "INTO",
+    "IS",
+    "LIKE",
+    "NOT",
+    "NULL",
+    "OF",
+    "ON",
+    "OR",
+    "ORDER",
+    "REVOKE",
+    "ROW",
+    "ROWS",
+    "SAMPLE",
+    "SELECT",
+    "SET",
+    "START",
+    "TABLE",
+    "TABLESAMPLE",
+    "THEN",
+    "TO",
+    "TRIGGER",
+    "UNION",
+    "UNIQUE",
+    "UPDATE",
+    "VALUES",
+    "WHENEVER",
+    "WHERE",
+    "WITH",
+
+    // oracle reserved KW
+    "INCREMENT",
+    "MINUS",
+
+    // snowflake reserved KW
+    "AGGREGATE",
+    "ILIKE",
+    "REGEXP",
+    "RLIKE",
+    "SOME",
+    "QUALIFY"
+  };
+
+  private static final Set<String> reservedKeywordSet =
+      new HashSet<>(Arrays.asList(reservedKeywords));
+
+  // column names that do not need quoting to match metadata
+  private static final Pattern unquotedIdentifierPattern =
+      Pattern.compile("[$][0-9]+|[_A-Z][_A-Z0-9]*[$]?[_A-Z0-9]*");
+
+  /**
+   * Format the column name, given with the user inserted row, to match column metadata, sent by
+   * server side.
+   */
+  static String formatColumnName(String columnName) {
+    Utils.assertStringNotNullOrEmpty("invalid column name", columnName);
+    // internal server side normalised column name in persistent metadata
+    String normalisedInternalName = unquoteColumnName(columnName);
+    // display name sent by server side to client in open channel column metadata response
+    return quoteColumnNameIfNeeded(normalisedInternalName);
+  }
+
+  /**
+   * Quote user column name to match server side metadata.
+   *
+   * <p>This needs to be in sync with server side to code, until we decide to send the normalised
+   * unquoted column name from server.
```

Comment:
Would it be better to do the unquoted column name instead rather than these changes?

Reply:
We need both quoted and unquoted - everything on REST API expects quoted and everything in BDEC expects unquoted (XP expects unquoted).

Reply:
And I need to update this comment - even if we send both quoted and unquoted from the server side, we still need quote user input to match the column display name.